bin/changelog/zenhub/requests.js

```diff
@@ -0,0 +1,101 @@
+'use strict';
+
+/* eslint no-console: 0 */
+
+const ZenHub = require( 'zenhub-api' );
+const { REPO } = require( '../config' );
+const { authedGraphql } = require( '../common' );
+const { pull } = require( 'lodash' );
+
+const api = new ZenHub( process.env.ZH_API_KEY );
+
+const getQuery = ( before ) => {
+	const [ owner, repo ] = REPO.split( '/' );
+	const paging = before ? `, before: "${ before }"` : '';
+	const query = `
+	{
+		repository(owner: "${ owner }", name: "${ repo }") {
+			pullRequests(last: 100, states: [MERGED], orderBy: {field: CREATED_AT, direction: ASC}${ paging }) {
+				totalCount
+				pageInfo {
+					hasNextPage
+					endCursor
+				}
+				nodes {
+					number
+					title
+					url
+					author {
+						login
+					}
+					body
+					labels(last: 10) {
+						nodes {
+							name
+						}
+					}
+				}
+			}
+		}
+	}
+	`;
+	return query;
+};
+
+const fetchAllIssuesForRelease = async ( releaseId ) => {
+	const releaseIssues = await api.getReleaseReportIssues( {
+		release_id: releaseId,
+	} );
+	return releaseIssues.map( ( releaseIssue ) => releaseIssue.issue_number );
+};
+
+const extractPullRequestsMatchingReleaseIssue = (
+	releaseIds,
+	pullRequests
+) => {
+	return pullRequests.filter( ( pullRequest ) => {
+		const hasPullRequest = releaseIds.includes( pullRequest.number );
+		if ( hasPullRequest ) {
+			pull( releaseIds, pullRequest.number );
+			return true;
+		}
+		return false;
+	} );
+};
+
+const fetchAllPullRequests = async ( releaseId ) => {
+	// first get all release issue ids
+	const releaseIds = await fetchAllIssuesForRelease( releaseId );
```

Comment:
this script works in the assumption that all PRs are recently merged and submitted and might not be able to account for older ones.

One other method that I have in my mind that I'm not sure how to implement is just to keep hitting the graphql endpoint until we covered all pulls, so imagine a release returning this:

```
const releaseIds = [
  1134,
  1135,
  1212,
  1213,
  1214
]
```

assuming all issues are closed and all PRs are merged, we would query the API for both closed issues and merged PRs, filter out issues from `releaseIds`, filter out matching PRs from `releaseIds`, filter out issues from the returned result, check if `releaseIds` length is bigger than 0, increase page number, do this again.

This seems like an overkill and sure entirely how to do it, but I can keep it in mind.

Reply:
Ya definitely something to consider if we find we're still missing pulls and was an approach I was thinking of doing initially.  However, I think until github provides more filtering options, even with the potential for older pulls etc we should be okay because:

- The request is already pre-filtering on _merged_ pulls. So that basically eliminates all issues and closed (but not merged) pulls from the results.
- The number of pages being queried is based on the release id results which _includes_ issues and closed pulls, so already the number of pages is likely greater than what's needed.
- I've added a couple pages as a buffer just in case.
- The results are 100 per page which is already a large buffer.

For example, for the latest release cycle, there are around 190 issues+pulls  in it.  That means there are 4 pages going to be queried or 400 results.  Since we're only retrieving _merged pulls_ from github, I think that's more than enough to catch all the pulls in that release.